sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_io/base.py

```diff
@@ -374,14 +380,57 @@ def __init__(
                 no_personal_data_message=msg.format("[data]"),
             )
         super().__init__(meta=meta, data=data, **kwargs)
-        self._name = name
+        self._port_name = port_name
         self._owner = owner
+        self._name = data.name if isinstance(data, Output) else None
+        self._version = data.version if isinstance(data, Output) else None
+
+        self._assert_name_and_version_exist()
+
         self._is_control = meta.is_control if meta is not None else None
 
     @property
     def is_control(self) -> str:
         return self._is_control
 
+    @property
+    def port_name(self) -> str:
+        return self._port_name
+
+    @property
+    def name(self):
+        return self._name
+
+    @name.setter
+    def name(self, name):
+        self._build_default_data()
+        self._name = name
+        if isinstance(self._data, (Input, Output)):
+            self._data.name = name
+        else:
+            self._data._name = name
+
+    @property
+    def version(self) -> str:
+        return self._version
+
+    @version.setter
+    def version(self, version):
+        self._build_default_data()
+        self._data.type = self.type
+        self._version = version
+        if isinstance(self._data, (Input, Output)):
+            self._data.version = version
+        else:
+            self._data._version = version
+
+    def _assert_name_and_version_exist(self):
+        if self.version and not self.name:
+            class_name = self.__class__.__name__
+            raise UserErrorException(
+                f"""{class_name}.name is missing. We don't support registering output only with version.
```

Comment:
May use a shorter description:

"Output name is required when output version is specified."